### PR TITLE
SRKN

### PR DESCRIPTION
**Pull request by eggrobin**
Split SRKN and SPRK, use SRKN just about everywhere.

### CODE REVIEW

File: benchmarks/symplectic_partitioned_runge_kutta_integrator.cpp [hunk 1]
```diff
@@ -142,32 +142,26 @@
...
-
-  integrator.Initialize((integrator.*scheme)());
+    not_null<Speed*> const v_error,
+    SRKNIntegrator const& integrator) {
+  SRKNIntegrator::Solution<Length, Speed> solution;
+  SRKNIntegrator::Parameters<Length, Speed> parameters;
 
   parameters.initial.positions.emplace_back(SIUnit<Length>());
-  parameters.initial.momenta.emplace_back(Momentum());
+  parameters.initial.momenta.emplace_back(Speed());
```
Comment (pleroy) [inline benchmarks/symplectic_partitioned_runge_kutta_integrator.cpp]: Do we need `Speed()`?  Could the `emplace` default-construct?

File: integrators/symplectic_integrator.hpp [hunk 2]
```diff
@@ -24,26 +24,31 @@ struct DoublePrecision {
   Scalar error;
 };
 
-template<typename Position, typename Momentum>
 class SymplecticIntegrator {
  public:
-  SymplecticIntegrator() = default;
   virtual ~SymplecticIntegrator() = default;
```
Comment (pleroy) [inline integrators/symplectic_integrator.hpp]: Destructor after constructor.

File: integrators/symplectic_runge_kutta_nystrom_integrator.hpp [hunk 3]
```diff
@@ -0,0 +1,124 @@
...
+      Parameters<Position, Variation<Position>> const& parameters,
+      not_null<Solution<Position, Variation<Position>>*> const solution) const;
+};
+
+// Fourth order, 4 stages.  This method minimizes the error constant.
+// Coefficients from Robert I. McLachlan and Pau Atela (1992),
+// The accuracy of symplectic integrators, table 2.
+// http://eaton.math.rpi.edu/CSUMS/Papers/Symplectic/McLachlan_Atela_92.pdf.
+SRKNIntegrator const& McLachlanAtela1992Order4Optimal();
+// Fifth order, 6 stages.  This method minimizes the error constant  Ibidem.
```
Comment (pleroy) [inline integrators/symplectic_runge_kutta_nystrom_integrator.hpp]: Period after constant.

File: integrators/symplectic_runge_kutta_nystrom_integrator.hpp [hunk 4]
```diff
@@ -0,0 +1,124 @@
...
+
+  VanishingCoefficients vanishing_coefficients_;
+  // Null if, and only if, |vanishing_coefficients_ == kNone|.
+  // If |vanishing_coefficients_ == kFirstBVanishes|, this contains the first
+  // and last a coefficients.
+  // If |vanishing_coefficients_ == FirstAVanishes|, this contains the first
+  // and last b coefficients.
+  // These are used to desynchronize and synchronize first-same-as-last
+  // integrators.
+  std::unique_ptr<FirstSameAsLast> first_same_as_last_;
```
Comment (pleroy) [inline integrators/symplectic_runge_kutta_nystrom_integrator.hpp]: Comment std::optional.

File: integrators/symplectic_runge_kutta_nystrom_integrator_body.hpp [hunk 5]
```diff
@@ -0,0 +1,326 @@
...
+                                           0.4423637942197494587},
+                                         { 0.1193900292875672758,
+                                           0.6989273703824752308,
+                                          -0.1713123582716007754,
+                                           0.4012695022513534480,
+                                           0.0107050818482359840,
+                                          -0.0589796254980311632});
+  return integrator;
+}
+
```
Comment (pleroy) [inline integrators/symplectic_runge_kutta_nystrom_integrator_body.hpp]: Remove.

File: integrators/symplectic_runge_kutta_nystrom_integrator_body.hpp [hunk 6]
```diff
@@ -0,0 +1,326 @@
...
+
+inline SRKNIntegrator::SRKNIntegrator(std::vector<double> const& a,
+                                      std::vector<double> const& b)
+    : a_(std::move(a)),
```
Comment (pleroy) [inline integrators/symplectic_runge_kutta_nystrom_integrator_body.hpp]: Don't move!

File: integrators/symplectic_runge_kutta_nystrom_integrator_body.hpp [hunk 7]
```diff
@@ -0,0 +1,326 @@
...
+void SRKNIntegrator::SolveTrivialKineticEnergyIncrementOptimized(
+    RightHandSideComputation compute_acceleration,
+    Parameters<Position, Variation<Position>> const& parameters,
+    not_null<Solution<Position, Variation<Position>>*> const solution) const {
+  using Velocity = Variation<Position>;
+  int const dimension = parameters.initial.positions.size();
+
+  std::vector<Position> Δqstage0(dimension);
+  std::vector<Position> Δqstage1(dimension);
+  std::vector<Velocity> Δpstage0(dimension);
```
Comment (pleroy) [inline integrators/symplectic_runge_kutta_nystrom_integrator_body.hpp]: Rename.

File: integrators/symplectic_runge_kutta_nystrom_integrator_body.hpp [hunk 8]
```diff
@@ -0,0 +1,326 @@
...
+  solution->clear();
+  solution->reserve(capacity);
+
+  std::vector<DoublePrecision<Position>> q_last(parameters.initial.positions);
+  std::vector<DoublePrecision<Velocity>> v_last(parameters.initial.momenta);
+  int sampling_phase = 0;
+
+  std::vector<Position> q_stage(dimension);
+  std::vector<Velocity> v_stage(dimension);
+  std::vector<Quotient<Velocity, Time>> a(dimension);  // Current forces.
```
Comment (pleroy) [inline integrators/symplectic_runge_kutta_nystrom_integrator_body.hpp]: acceleration

File: integrators/simple_harmonic_motion.cpp [hunk 9]
```diff
@@ -0,0 +1,363 @@
...
+// "where GetParam() = Leapfrog" rather than
+// "where GetParam() = n-byte object <hex>"
+std::ostream& operator<<(std::ostream& stream,
+                         SimpleHarmonicMotionTestableProperties param) {
+  return stream << param.name;
+}
+
+std::vector<SimpleHarmonicMotionTestableProperties> Instances() {
+  return {
+      {INTEGRATOR(Leapfrog), 2, 0.4 * Second,
```
Comment (pleroy) [inline integrators/simple_harmonic_motion.cpp]: I'd prefer a line break systematically after `INTEGRATOR(...),` to make sure that all entries look the same.

File: integrators/simple_harmonic_motion.cpp [hunk 10]
```diff
@@ -0,0 +1,363 @@
...
+    : public testing::TestWithParam<SimpleHarmonicMotionTestableProperties> {
+ public:
+  static void SetUpTestCase() {
+    google::LogToStderr();
+  }
+
+ protected:
+  SimpleHarmonicMotionTest() : integrator_(GetParam().integrator) {}
+
+  not_null<SRKNIntegrator const*> const     integrator_;
```
Comment (pleroy) [inline integrators/simple_harmonic_motion.cpp]: We don't normally indent identifiers this way.

File: integrators/simple_harmonic_motion.cpp [hunk 11]
```diff
@@ -0,0 +1,363 @@
...
+}
+
+TEST_P(SimpleHarmonicMotionTest, Convergence) {
+  parameters_.initial.positions.emplace_back(SIUnit<Length>());
+  parameters_.initial.momenta.emplace_back(Speed());
+  parameters_.initial.time = Time();
+  parameters_.tmax = 100 * SIUnit<Time>();
+  parameters_.sampling_period = 0;
+  // For 0.2 * 1.1⁻²¹ < |Δt| < 0.2 , the correlation between step size and error
+  // is very strong. It the step is small enough to converge and large enough to
```
Comment (pleroy) [inline integrators/simple_harmonic_motion.cpp]: This sentence does not parse.

File: integrators/symplectic_partitioned_runge_kutta_integrator_test.cpp [hunk 12]
```diff
@@ -50,127 +51,57 @@ using ::testing::Lt;
 using ::testing::Ne;
 using ::testing::ValuesIn;
 
-#define INTEGRATOR(name) &Integrator::name, #name
+#define INTEGRATOR(name) &name(), #name
```
Comment (pleroy) [inline integrators/symplectic_partitioned_runge_kutta_integrator_test.cpp]: Is there an #undef somewhere (I couldn't find it but who knows...).

File: integrators/symplectic_runge_kutta_nystrom_integrator.hpp [hunk 13]
```diff
@@ -0,0 +1,124 @@
...
+    kFirstBVanishes,
+    kLastAVanishes,
+  };
+
+  struct FirstSameAsLast {
+    double first;
+    double last;
+  };
+
+  VanishingCoefficients vanishing_coefficients_;
```
Comment (pleroy) [inline integrators/symplectic_runge_kutta_nystrom_integrator.hpp]: Comment this member.

File: integrators/symplectic_runge_kutta_nystrom_integrator_test.cpp [hunk 14]
```diff
@@ -0,0 +1,215 @@
...
+using testing_utilities::RelativeError;
+using testing_utilities::Slope;
+using ::testing::AllOf;
+using ::testing::Eq;
+using ::testing::Gt;
+using ::testing::Lt;
+using ::testing::Ne;
+using ::testing::ValuesIn;
+
+#define INTEGRATOR(name) &name(), #name
```
Comment (pleroy) [inline integrators/symplectic_runge_kutta_nystrom_integrator_test.cpp]: #undef somewhere.

File: benchmarks/symplectic_partitioned_runge_kutta_integrator.cpp (hunk 1 above)
Comment (eggrobin) [inline benchmarks/symplectic_partitioned_runge_kutta_integrator.cpp]: It's clearer like this.

File: integrators/simple_harmonic_motion.cpp (hunk 9 above)
Comment (eggrobin) [inline integrators/simple_harmonic_motion.cpp]: Breaking after the order instead.

File: integrators/simple_harmonic_motion.cpp (hunk 10 above)
Comment (eggrobin) [inline integrators/simple_harmonic_motion.cpp]: Done.

File: integrators/simple_harmonic_motion.cpp (hunk 11 above)
Comment (eggrobin) [inline integrators/simple_harmonic_motion.cpp]: That whole comment is obsolete now that we have `beginning_of_convergence` actually.

File: integrators/symplectic_partitioned_runge_kutta_integrator_test.cpp (hunk 12 above)
Comment (eggrobin) [inline integrators/symplectic_partitioned_runge_kutta_integrator_test.cpp]: It's a cpp.

File: integrators/symplectic_runge_kutta_nystrom_integrator.hpp (hunk 13 above)
Comment (eggrobin) [inline integrators/symplectic_runge_kutta_nystrom_integrator.hpp]: Done.

File: integrators/symplectic_runge_kutta_nystrom_integrator.hpp (hunk 4 above)
Comment (eggrobin) [inline integrators/symplectic_runge_kutta_nystrom_integrator.hpp]: Done.

File: integrators/symplectic_runge_kutta_nystrom_integrator.hpp (hunk 3 above)
Comment (eggrobin) [inline integrators/symplectic_runge_kutta_nystrom_integrator.hpp]: Done.

File: integrators/symplectic_runge_kutta_nystrom_integrator_body.hpp (hunk 5 above)
Comment (eggrobin) [inline integrators/symplectic_runge_kutta_nystrom_integrator_body.hpp]: Done.

File: integrators/symplectic_runge_kutta_nystrom_integrator_body.hpp (hunk 6 above)
Comment (eggrobin) [inline integrators/symplectic_runge_kutta_nystrom_integrator_body.hpp]: Done.

File: integrators/symplectic_runge_kutta_nystrom_integrator_body.hpp (hunk 7 above)
Comment (eggrobin) [inline integrators/symplectic_runge_kutta_nystrom_integrator_body.hpp]: Done.

File: integrators/symplectic_runge_kutta_nystrom_integrator_body.hpp (hunk 8 above)
Comment (eggrobin) [inline integrators/symplectic_runge_kutta_nystrom_integrator_body.hpp]: Done. Also removed `v`.

File: integrators/symplectic_runge_kutta_nystrom_integrator_test.cpp (hunk 14 above)
Comment (eggrobin) [inline integrators/symplectic_runge_kutta_nystrom_integrator_test.cpp]: It's a cpp.

**Pull request merged**